File: examples/example_cpu.py
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image

from mesh_utils import Mesh, CPUBuilder, CPUTraverser


# ==== Load and prepare BVH ==== #

mesh = Mesh.from_file("suzanne.fbx")
# mesh.split_faces(0.9)

builder = CPUBuilder(mesh)
bvh_data = builder.build_bvh(25)
bvh_data.save_to_obj("bvh.obj", 25)
bvh = CPUTraverser(bvh_data)

img_size = 800
n_pixels = img_size * img_size


# ==== Generate rays ==== #

mesh_min, mesh_max = mesh.get_bounds()
max_extent = max(mesh_max - mesh_min)

center = (mesh_max + mesh_min) * 0.5

cam_pos = np.array([
    center[0] + max_extent * 1.0,
    center[1] - max_extent * 1.5,
    center[2] + max_extent * 0.5,
])
cam_poses = np.tile(cam_pos, (n_pixels, 1))
cam_dir = (center - cam_pos) * 0.9

x_dir = np.cross(cam_dir, np.array([0, 0, 1]))
x_dir = x_dir / np.linalg.norm(x_dir) * (max_extent / 2)

y_dir = -np.cross(x_dir, cam_dir)
y_dir = y_dir / np.linalg.norm(y_dir) * (max_extent / 2)

x_coords, y_coords = np.meshgrid(
    np.linspace(-1, 1, img_size),
    np.linspace(-1, 1, img_size),
)

x_coords = x_coords.flatten()
y_coords = y_coords.flatten()

dirs = cam_dir[None, :] + x_dir[None, :] * x_coords[:, None] + y_dir[None, :] * y_coords[:, None]


# ==== Run BVH ==== #

depths = np.zeros((cam_poses.shape[0],), dtype=np.uint32)
prim_idxs = np.zeros((cam_poses.shape[0],), dtype=np.uint32)
mask = np.zeros((cam_poses.shape[0],), dtype=np.bool_)
t = np.zeros((cam_poses.shape[0],), dtype=np.float32) + 1e9
normals = np.zeros((cam_poses.shape[0], 3), dtype=np.float32)
uvs = np.zeros((cam_poses.shape[0], 2), dtype=np.float32)

bvh.ray_query(cam_poses, dirs, mask, t, prim_idxs, normals, uvs)

# ==== Visualize ==== #

mask_img = mask.reshape(img_size, img_size)

light_dir = np.array([1, -1, 1])
light_dir = light_dir / np.linalg.norm(light_dir)

normals[np.isnan(normals)] = 0
colors = np.dot(normals, light_dir) * 0.5 + 0.5

img = colors.reshape(img_size, img_size)
img[~mask_img] = 0

image = Image.fromarray((img * 255).astype(np.uint8))
image.save('output.png')

File: examples/example_cpu.cpp
#include "../src/cpu_traverse.h"

int main() {
    /* ==== Load and prepare BVH ==== */

    cout << "Loading scene..." << endl;
    Mesh mesh("models/lego.fbx");
    cout << "Number of vertices: " << mesh.vertices.size() << endl;
    cout << "Number of faces: " << mesh.faces.size() << endl;

    cout << endl;

    cout << "Splitting faces..." << endl;
    // mesh.split_faces(0.9);
    cout << "Number of vertices: " << mesh.vertices.size() << endl;
    cout << "Number of faces: " << mesh.faces.size() << endl;

    cout << endl;

    CPUBuilder builder(mesh);
    cout << "Building BVH..." << endl;
    timer_start();
    BVHData bvh_data = builder.build_bvh(5);
    cout << "Elapsed time: " << timer_stop() << " ms" << endl;
    cout << "Number of nodes: " << bvh_data.n_nodes << endl;
    cout << "Number of leaves: " << bvh_data.n_leaves << endl;
    cout << "Depth: " << bvh_data.depth << endl;
    bvh_data.save_as_obj("bvh.obj", 12);
    CPUTraverser bvh(bvh_data);

    cout << endl;

    /* ==== Setting up camera ==== */

    auto [min, max] = mesh.bounds();
    glm::vec3 center = (max + min) * 0.5f;
    float max_extent = std::fmax(max.x - min.x, std::fmax(max.y - min.y, max.z - min.z));
    glm::vec3 cam_pos = { 
        center.x + max_extent * 1.0,
        center.y - max_extent * 1.5,
        center.z + max_extent * 0.5
    };
    glm::vec3 cam_dir = (center - cam_pos) * 0.9f;
    glm::vec3 x_dir = glm::normalize(glm::cross(cam_dir, glm::vec3(0, 0, 1))) * (max_extent / 2);
    glm::vec3 y_dir = -glm::normalize(glm::cross(x_dir, cam_dir)) * (max_extent / 2);
    int img_size = 800;
    std::vector<glm::vec3> img(img_size * img_size, glm::vec3(0));
    glm::vec3 light_dir = glm::normalize(glm::vec3(1, -1, 1));


    /* ==== Rendering image ==== */

    cout << "Rendering image..." << endl;
    timer_start();
    // #pragma omp parallel for
    for (int y = 0; y < img_size; y++) {
        for (int x = 0; x < img_size; x++) {
            float x_f = ((float)x / img_size - 0.5f) * 2;
            float y_f = ((float)y / img_size - 0.5f) * 2;

            glm::vec3 dir = cam_dir + x_dir * x_f + y_dir * y_f;
            HitResult hit = bvh.closest_primitive_single({cam_pos, dir});

            if (hit.hit) {
                float att = glm::dot(light_dir, hit.normal) * 0.5 + 0.5;
                glm::vec3 color = hit.color * att;
                img[y * img_size + x] = color;
            }
        }
    }
    cout << "Elapsed time: " << timer_stop() << " ms" << endl;

    cout << endl;


    /* ==== Saving image ==== */

    std::vector<unsigned int> pixels(img_size * img_size);
    for (int i = 0; i < img_size * img_size; i++) {
        pixels[i] = (255 << 24) | ((int)(img[i].z * 255) << 16) | ((int)(img[i].y * 255) << 8) | (int)(img[i].x * 255);
    }
    cout << "Saving image..." << endl;
    save_to_bmp(pixels.data(), img_size, img_size, "output.bmp");

    return 0;
}
